plasmapy/analysis/swept_langmuir/floating_potential.py

```diff
@@ -0,0 +1,295 @@
+        documentation of the `rsq` property on the associated fit function
+        (e.g. the
+        `~plasmapy.analysis.fit_functions.ExponentialPlusOffset.rsq`
+        property of
+        `~plasmapy.analysis.fit_functions.ExponentialPlusOffset`).
+
+    func: sub-class of `~plasmapy.analysis.fit_functions.AbstractFitFunction`
+        The callable function :math:`f(x)` representing the fit and its results.
+
+    islands: `List[slice]`
+        List of `slice` objects representing the indices of the identified
+        crossing-islands.
+
+    indices: `slice`
+        A `slice` object representing the indices of ``voltage`` and ``current``
+        arrays used for the fit.
+
+    """
+    rtn = FloatingPotentialResults(
+        vf=np.nan, vf_err=np.nan, rsq=None, func=None, islands=None, indices=None
+    )._asdict()
+
+    _settings = {
+        "linear": {"func": ffuncs.Linear, "min_point_factor": 0.1},
+        "exponential": {"func": ffuncs.ExponentialPlusOffset, "min_point_factor": 0.2},
+    }
+    try:
+        min_point_factor = _settings[fit_type]["min_point_factor"]
+        fit_func = _settings[fit_type]["func"]()
+        rtn["func"] = fit_func
+    except KeyError:
+        raise KeyError(
+            f"Requested fit function '{fit_type}' is  not a valid option.  "
+            f"Examine kwarg 'fit_curve' for valid options."
+        )
+
+    # check voltage and current arrays
+    try:
+        check_sweep(voltage, current)
+    except (TypeError, ValueError) as err:
+        # did not pass checks
+        warn(err.args[0], PlasmaPyWarning)
+        return FloatingPotentialResults(**rtn)
+
+    # condition kwarg threshold
+    if isinstance(threshold, (int, np.integer)):
+        if threshold < 1:
+            threshold = 1
+            warn(
+                f"threshold ({threshold}) is less than 1 and needs to be"
+                f" an int >= 1, using a value of 1",
+                PlasmaPyWarning,
+            )
+
+    else:
+        threshold = 1
+        warn(f"threshold is NOT a integer >= 1, using a value of 1", PlasmaPyWarning)
+
+    # condition min_points
+    if min_points is None:
+        min_points = int(np.max([5, np.around(min_point_factor * voltage.size)]))
+    elif min_points == 0:
+        # this signals to use all points
+        pass
+    elif 0 < min_points < 1:
+        min_points = int(np.round(min_points * voltage.size))
+    elif min_points >= 1:
+        min_points = int(np.round(min_points))
+    else:
+        raise ValueError(
+            f"Got {min_points}, but 'min_points' must be an int or float "
+            f"greater than or equal to 0."
+        )
+
+    # find possible crossing points (cp)
+    lower_vals = current < 0
+    upper_vals = 0 < current
+    cp_exact = (current == 0.0).nonzero()[0]
+    cp_low2high = np.logical_and(np.roll(lower_vals, 1), upper_vals).nonzero()[0]
+    cp_high2low = np.logical_and(np.roll(lower_vals, -1), upper_vals).nonzero()[0]
```

Comment:
While I'm usually supportive of this idea, I don't think the code here is complex or long enough to warrant this approach.  I feel the "extract method refactoring` reduces readability when the code is not complex or long, since the user/programmer has to woner away from the current line to find out what the method does.

Reply:
Huh...I've lately been thinking that 20-line functions are long!  I'll still encourage this, but let you make the final decision.

Reply:
Oh...I don't really consider the length/complexity of the "parent" function when I break off the method, but more the length/complexity of the code that would go into the "method".  If the "method" code easily fits within 2/3 of my screen (i.e. I can still easily read the pre- and post-sections around the "method" code) then I leave it in place.  If I lose the readability of the pre- and post-sections around the "method" code, then I'll break that code-block off into it's own method.  I do it this way, because I find when the code is broken up too much then it's really hard for me to follow it, even when the method names are "explicit."